### PR TITLE
Spectrum to allow spec timeout to be specified throughout the hierarchy

### PR DESCRIPTION
**Pull request by ashleyfrieze**
An implementation to address #74 

Still WIP at the moment, feedback appreciated.

### CODE REVIEW

File: docs/Configuration.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+## Configuration
```
Comment (ashleyfrieze) [inline docs/Configuration.md]: This wasn't incredibly well documented.

File: docs/FocusingAndIgnoring.md [hunk 2]
```diff
@@ -100,8 +100,18 @@ describe("Ignored specs", () -> {
 ```
 #### Ignoring Specs Spectrum style
 
-Spectrum allows you to specify preconditions on a block. These preconditions can include tagging.
-The `with(BlockConfigurationChain,Block)` function is used to annotate a block with preconditions. Preconditions include:
+Spectrum allows you to specify preconditions for execution as part of the configuration of a block. These can include tagging.
```
Comment (ashleyfrieze) [inline docs/FocusingAndIgnoring.md]: Removing `precondition` as a keyword in relation to the block metadata, which we're really calling `BlockConfiguration`

File: docs/README.md [hunk 3]
```diff
@@ -25,4 +25,4 @@ Unlike some BDD-style frameworks, Spectrum is _only_ a test runner. Assertions,
 
 ## Getting Started
 
-See the [quickstart waklthrough](QuickstartWalkthrough.md) and the docs for writing [Specification-style](SpecificationDSL.md) or [Gherkin-style](GherkinDSL.md) tests.
+See the [quickstart walkthrough](QuickstartWalkthrough.md) and the docs for writing [Specification-style](SpecificationDSL.md) or [Gherkin-style](GherkinDSL.md) tests.
```
Comment (ashleyfrieze) [inline docs/README.md]: A _tyop_

File: src/main/java/com/greghaskins/spectrum/Configure.java [hunk 4]
```diff
@@ -96,6 +100,27 @@ static BlockConfigurationChain focus() {
...
+   * own timeout.
+   * @param timeout the amount of the timeout
+   * @param timeUnit the unit of the timeout
+   * @return a chainable configuration that will apply a timeout to all leaf nodes below
+   */
+  static BlockConfigurationChain timeout(long timeout, TimeUnit timeUnit) {
+    return new BlockConfigurationChain().with(new BlockTimeout(timeout, timeUnit));
+  }
+
+  /**
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/Configure.java]: Timeouts. We waited long enough for them.

File: src/main/java/com/greghaskins/spectrum/internal/Child.java [hunk 5]
```diff
@@ -40,19 +41,4 @@ default boolean isAtomic() {
...
-   * @param taggingFilterCriteria the tagging state in the parent of this suite or spec.
-   *                     This is used to determine what filters apply to the block
-   * @return the child itself for fluent calling
-   */
-  default Child applyPreconditions(final Block block,
-      final TaggingFilterCriteria taggingFilterCriteria) {
-    ConfiguredBlock.findApplicablePreconditions(block).applyTo(this, taggingFilterCriteria);
-
-    return this;
-  }
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/Child.java]: This didn't have to be here. Its implementation has moved.

File: src/main/java/com/greghaskins/spectrum/internal/LeafChild.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.greghaskins.spectrum.internal;
+
+import com.greghaskins.spectrum.internal.hooks.Hook;
+
+/**
+ * Interface for a child that is also a leaf node in the test hierarchy.
+ */
+public interface LeafChild extends Child {
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/LeafChild.java]: This _leaf_ interface allows us to use `instanceof` and a cast to specific type in order to work with a leaf node to apply leaf-level hooks. If we didn't do that we'd end up adding an unreachable method to `Child` for some of its descendants - the code coverage led me to this approach.

File: src/main/java/com/greghaskins/spectrum/internal/Spec.java [hunk 7]
```diff
@@ -1,18 +1,23 @@
...
+import com.greghaskins.spectrum.internal.hooks.Hooks;
 
 import org.junit.runner.Description;
 import org.junit.runner.notification.Failure;
 
-final class Spec implements Child {
+final class Spec implements LeafChild {
 
-  private final NotifyingBlock block;
+  private final Block block;
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/Spec.java]: The block is kept here in its raw form, as the leaf hooks will wrap it with reporting.

File: src/main/java/com/greghaskins/spectrum/internal/Spec.java [hunk 8]
```diff
@@ -31,8 +36,8 @@ public void run(final RunReporting<Description, Failure> notifier) {
       return;
     }
 
-    this.block.run(this.description, notifier);
-
+    // apply leaf hooks around the inner block
+    leafHooks.runAround(this.description, notifier, block);
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/Spec.java]: Leaf hooks now always run, even when there are none.

File: src/main/java/com/greghaskins/spectrum/internal/Suite.java [hunk 9]
```diff
@@ -30,7 +31,7 @@
   private boolean ignored;
 
   private final TaggingFilterCriteria tagging;
-  private BlockConfiguration preconditions = BlockConfiguration.defaultConfiguration();
+  private BlockConfiguration configuration = BlockConfiguration.defaultConfiguration();
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/Suite.java]: Reads better after rename

File: src/main/java/com/greghaskins/spectrum/internal/Suite.java [hunk 10]
```diff
@@ -70,9 +71,10 @@ public Suite addSuite(final String name) {
   }
 
   private Suite addSuite(final String name, final ChildRunner childRunner) {
-    final Suite suite =
-        new Suite(Description.createSuiteDescription(sanitise(name)), this, childRunner,
-            this.tagging.clone());
+    final Suite suite = new Suite(Description.createSuiteDescription(sanitise(name)), this, childRunner,
+        this.tagging.clone());
+
+    suite.inheritConfigurationFromParent(configuration.forChild());
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/Suite.java]: We were not propagating configurations especially cleanly. It's now worked in.

File: src/main/java/com/greghaskins/spectrum/internal/configuration/BlockConfiguration.java [hunk 11]
```diff
@@ -12,12 +12,12 @@
...
   /**
-   * Combine provided precondition objects together.
+   * Combine provided configuration objects together.
    *
    * @param conditions to combine
-   * @return a combination of all preconditions as a new object
+   * @return a combination of all configurations as a new object
    */
-  static BlockConfiguration merge(BlockConfiguration... conditions) {
+  public static BlockConfiguration merge(BlockConfiguration... conditions) {
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/configuration/BlockConfiguration.java]: Opening this up seemed to make sense.

File: src/main/java/com/greghaskins/spectrum/internal/configuration/BlockTimeout.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    // my timeout supersedes any inherited timeout
+
+    return this;
+  }
+
+  private Hook timeoutHook() {
+    return withUnreportedInnerBlock((description, reporting, block) -> {
+      FailOnTimeout.builder().withTimeout(timeout, timeUnit)
+          .build(statementOf(block))
+          .evaluate();
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/configuration/BlockTimeout.java]: Essentially this is `JUnit`'s own implementation that I'm using here... it caused a few glitches, hence some reworking of the way that `NotifyingBlock` is used inside tests, since the multi-threaded nature of the above mechanism gave me some `InterruptedException` headaches.

File: src/main/java/com/greghaskins/spectrum/internal/configuration/TaggingFilterCriteria.java [hunk 13]
```diff
@@ -63,15 +63,10 @@ private boolean compliesWithRequired(Collection<String> tags) {
...
 
-  private String[] fromSystemProperty(final String property) {
+  private Stream<String> fromSystemProperty(final String property) {
     return Optional.ofNullable(System.getProperty(property))
-        .map(string -> string.split(TaggingFilterCriteria.TAGS_SEPARATOR))
-        .filter(TaggingFilterCriteria::notArrayWithEmptyValue)
-        .orElse(new String[0]);
+        .map(string -> Arrays.stream(string.split(TaggingFilterCriteria.TAGS_SEPARATOR))
+            .filter(tag -> !tag.isEmpty()))
+        .orElse(Stream.empty());
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/configuration/TaggingFilterCriteria.java]: Simplified in an attempt to improve code coverage.

File: src/main/java/com/greghaskins/spectrum/internal/hooks/Hook.java [hunk 14]
```diff
@@ -27,11 +27,40 @@ void accept(final Description description, final RunReporting<Description, Failu
       final Block block) throws Throwable;
 
   /**
+   * Override to return true if the inner block cannot report its own errors for some reason.
+   * @return true to suppress wrapping inner block in self-reporting
+   */
+  default boolean requiresUnreportedInnerBlock() {
+    return false;
+  }
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/hooks/Hook.java]: Hooks were being heavily wrapped in exception handling/reporting, but this is not necessary for all hooks.

File: src/main/java/com/greghaskins/spectrum/internal/hooks/Hooks.java [hunk 15]
```diff
@@ -95,25 +96,34 @@ private Hook createChainOfResponsibility(Variable<Boolean> hooksRememberedToRunT
...
-        wrapWithReporting(description, reporting,
+        conditionallyWrapWithReporting(inner, description, reporting,
             () -> inner.accept(description, reporting, block)));
   }
 
+  private static Block conditionallyWrapWithReporting(final Hook forHook, final Description description,
+      final RunReporting<Description, Failure> reporting, final Block innerBlock) {
+    if (forHook.requiresUnreportedInnerBlock()) {
+      return innerBlock;
+    }
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/hooks/Hooks.java]: This is kind of a workaround and kind of an efficiency - some hooks can ask for no wrapping with additional notification/error handling.

File: src/test/java/specs/JUnitRuleExample.java [hunk 16]
```diff
@@ -27,6 +27,16 @@
...
+  // alternative morphology of providing a rule - see http://junit.org/junit4/javadoc/4.12/org/junit/Rule.html
+  public static class TempFolderRuleProvidedViaMethodMixin {
+    private TemporaryFolder tempFolderRule = new TemporaryFolder();
+
+    @Rule
+    public TemporaryFolder getFolder() {
+      return tempFolderRule;
+    }
+  }
+
```
Comment (ashleyfrieze) [inline src/test/java/specs/JUnitRuleExample.java]: This permutation of JUnit `@Rule` wasn't tested before, though supported in the code.

File: src/test/java/specs/TimeoutSpecs.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.TimeUnit;
+
+@RunWith(Spectrum.class)
+public class TimeoutSpecs {
+  {
+    describe("A suite with timeouts", () -> {
+      it("will allow things to pass if they run quicker than the timeout", () -> {
+        final Result result = SpectrumHelper.run(() -> {
```
Comment (ashleyfrieze) [inline src/test/java/specs/TimeoutSpecs.java]: Using this approach rather than method-level classes, as I think it reads better... feedback appreciated.

**Comment by ashleyfrieze** [comment]
@greghaskins - I'm hoping you'll find this an easy PR to approve, given I'm just following the `with`/`and` stuff we set up for `ignore`/`focus` and `tags`.

**Comment by ashleyfrieze** [comment]
One TODO left - I want to make the `Hooks` at _leaf_ level unable to do weird stuff with `timeout`. Need to decide how to do that. Probably limit leaf hooks from requiring error notification/reporting.

File: src/test/java/specs/TimeoutSpecs.java (hunk 17 above)
Comment (greghaskins) [reply src/test/java/specs/TimeoutSpecs.java]: Agreed. I added that test helper for this exact purpose, as I got sick of writing nested classes.

File: src/test/java/specs/TimeoutSpecs.java [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+          describe("Suite with low timeout", with(timeout(1, TimeUnit.MILLISECONDS), () -> {
+            describe("with child suite", () -> {
+              it("has a more generous timeout", with(timeout(1, TimeUnit.MINUTES), () -> {
+                Thread.sleep(100);
+              }));
+            });
+          }));
+        });
+        assertThat(result.getFailureCount(), is(0));
```
Comment (greghaskins) [inline src/test/java/specs/TimeoutSpecs.java]: I noticed all the assertions here are confirming that tests fail/pass in the right scenarios. What about the messaging and user feedback that goes along with that? I haven't reviewed the implementation yet, but I assume there are some helpful messages that show up when one of these fail ("Timeout expired" or something to that effect).

**Comment by greghaskins** [comment]
I very much like this feature. 👍 Especially for being able to use a specified time unit.

Question: why not use the [Java 8 Duration](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html) object? `Duration.ofSeconds(5)` is pretty nice, which could be further simplified like `with(timeout(ofSeconds(5))) ...` after importing things statically.

I, for one, always lose time trying to count zeros and mentally convert between time units in my tests, so I almost feel like we should make units mandatory for readability. (Using a`Duration` or just making that `TimeUnit` parameter required would accomplish that.)

File: src/test/java/specs/TimeoutSpecs.java (hunk 18 above)
Comment (greghaskins) [reply src/test/java/specs/TimeoutSpecs.java]: Ah, I see it's leveraging JUnit4's [`FailOnTimeout` helper](https://github.com/junit-team/junit4/blob/master/src/main/java/org/junit/internal/runners/statements/FailOnTimeout.java). Might still be worth verifying some of that behavior in case we end up switching to something else down the line (JUnit5 might have a different helper, or we write our own).

Comment (ashleyfrieze) [reply src/test/java/specs/TimeoutSpecs.java]: We could wrap the helper's exception with a Spectrum one and then verify that, giving a seamless switch between implementations, or we could not promise anything special beyond _it fails_. What do you think?

**Comment by ashleyfrieze** [comment]
For full polyglot-like _it is how you are expecting_ for JUnit aficionados, the pure millisecond version is most natural, followed by the two parameter version as used in the timeout rule. I really like your Duration version, though. Could I just add it as an overload?

**Comment by ashleyfrieze** [comment]
Note: there will be a squash and rebase. Other than that, I'm pretty much ready with this, beyond feedback driven changes.

**Comment by ashleyfrieze** [comment]
@greghaskins - here is why I think we should NOT implement `Duration` in `timeout`, even though I think it's a good idea in theory.

1. The low-level implementation has to be turned into either _milliseconds_ or `long` + `TimeUnit` to be compatible with the Java thread waiting mechanism
2. `Duration` is weirdly in _seconds_ and _nanoseconds of a second_ which creates some awkward calculation
3. To test the calculation behaviourally with actual timeouts would suck so...
4. We could make timeout setting use interfaces JUST because of `Duration` and test with mocks, but that would suck
5. We could make the timeout conversion from `Duration` convert to an object with a `long` and a `TimeUnit` in it, so we can unit test the POJO... and... 
6. We'd have just created a custom `Duration` class of our own that we're not even going to let people use!

It seems a bit futile. In general, Java waits are expressed in milliseconds, or in `long` + `TimeUnit`. See `System.currentTimeMillis` and `Object#wait` and `Thread.sleep`.

File: src/main/java/com/greghaskins/spectrum/internal/junit/TimeoutWrapper.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.junit.internal.runners.statements.FailOnTimeout;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wrap JUnit's timeout mechanism as a {@link NonReportingHook}.
+ */
+public interface TimeoutWrapper {
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/junit/TimeoutWrapper.java]: It kind of made sense to have the `BlockTimeout` not do its own implementation of timeout and to move our dependency on JUnit4 into the `junit` package. This means we could implement the timeout hook another way in future and only change the factory method used... probably a sensible bit of design.

File: src/main/java/com/greghaskins/spectrum/internal/junit/TimeoutWrapper.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  /**
+   * Apply a timeout expressed as a duration to a builder of a {@link FailOnTimeout} object.
+   * @param builder to modify
+   * @param timeout duration of the timeout
+   * @return the builder input - for fluent use.
+   */
+  static FailOnTimeout.Builder withAppliedTimeout(FailOnTimeout.Builder builder, Duration timeout) {
+    builder.withTimeout(timeout.toNanos(), TimeUnit.NANOSECONDS);
```
Comment (ashleyfrieze) [inline src/main/java/com/greghaskins/spectrum/internal/junit/TimeoutWrapper.java]: While I had reservations about do this this way, it works and is simple. It _could_ be made more clever with different sorts of time range, and this function is easily tested, but it's not necessary.

File: src/test/java/specs/JUnitRuleExample.java [hunk 21]
```diff
@@ -77,6 +87,15 @@ public static void beforeClass() {
...
     });
+
+    describe("A spec with a rule mix-in where the rule is provided by method", () -> {
+      Supplier<TempFolderRuleProvidedViaMethodMixin> tempFolderRuleMixin =
+          junitMixin(TempFolderRuleProvidedViaMethodMixin.class);
+
+      it("has access to an initialised object", () -> {
+        assertNotNull(tempFolderRuleMixin.get().getFolder().getRoot());
+      });
+    });
```
Comment (ashleyfrieze) [inline src/test/java/specs/JUnitRuleExample.java]: Another missing bit of test

File: src/test/java/specs/TimeoutSpecs.java (hunk 17 above)
Comment (ashleyfrieze) [reply src/test/java/specs/TimeoutSpecs.java]: The nested classes were a rather surprising bit of java syntax - cleverly found!

File: src/test/java/specs/TimeoutSpecs.java (hunk 18 above)
Comment (ashleyfrieze) [reply src/test/java/specs/TimeoutSpecs.java]: I'm inclined not to worry about the exact failure - only that it fails.

**Comment by ashleyfrieze** [comment]
> @greghaskins - here is why I think we should NOT implement Duration in timeout, even though I think it's a good idea in theory.

Ok, so it turned out that, after I gave it some more thought, I changed my mind. This was because:

1. I want @greghaskins to have every reason to accept this PR quickly so I can do other stuff
2. I trusted an easy way to convert from `Duration` to `long`+`TimeUnit` via converting everything to nanoseconds
3. I thought the separation of concerns between Spectrum with `Duration` and `JUnit`'s internals in `long`+`TimeUnit` was good for ~me~ us.

I think the result is better.

**Comment by greghaskins** [comment]
Looks good. Merging.

A few edge cases come to mind that I didn't see tests for and haven't tried yet myself; curious how they would be handled:

1. What if the slowdown was in a `beforeEach` /`afterEach` block? I would expect this looks the same as if the test body itself was slow.
2. Where does the time for `beforeAll`/ `afterAll` count? For example, if my timeout is 5 seconds and `beforeAll` takes 7 seconds, what happens? Then, what about multiple and nested `beforeAll`/ `afterAll` blocks?
4. Do `aroundEach` and `aroundAll` work with this? I'm guessing yes.
3. Does the `timeout` hook interrupt long-running tests, or just fail if they go too long?

**Comment by ashleyfrieze** [comment]
@greghaskins - let's continue this discussion under #74
